File: 3. Behavioral Design Patterns/Observer/Go/main.go
package main

import "fmt"

// Observer Pattern Example in Go
// WeatherStation notifies Display observers when temperature changes

type Observer interface{ Update(float32) }

type Subject interface{ Attach(Observer); Detach(Observer); Notify() }

type WeatherStation struct{ observers []Observer; temperature float32 }
func (w *WeatherStation) Attach(o Observer){ w.observers = append(w.observers, o) }
func (w *WeatherStation) Detach(o Observer){ for i, obs := range w.observers { if obs == o { w.observers = append(w.observers[:i], w.observers[i+1:]...); break } } }
func (w *WeatherStation) Notify(){ for _, o := range w.observers { o.Update(w.temperature) } }
func (w *WeatherStation) SetTemperature(temp float32){ w.temperature = temp; w.Notify() }

type PhoneDisplay struct{}
func (PhoneDisplay) Update(temp float32){ fmt.Printf("Phone: Temperature is %.1f°C\n", temp) }

type WindowDisplay struct{}
func (WindowDisplay) Update(temp float32){ fmt.Printf("Window: Temperature is %.1f°C\n", temp) }

func main(){
	station := &WeatherStation{}
	phone := PhoneDisplay{}
	window := WindowDisplay{}
	station.Attach(phone)
	station.Attach(window)
	station.SetTemperature(25.5)
	station.SetTemperature(30.2)
}

File: 3. Behavioral Design Patterns/Observer/Rust/main.rs
// Observer Pattern Example in Rust
// WeatherStation notifies Display observers when temperature changes

trait Observer {
    fn update(&self, temperature: f32);
}

struct WeatherStation {
    observers: Vec<Box<dyn Observer>>,
    temperature: f32,
}

impl WeatherStation {
    fn new() -> Self { Self { observers: Vec::new(), temperature: 0.0 } }
    fn attach(&mut self, o: Box<dyn Observer>) { self.observers.push(o) }
    fn notify(&self) { for o in &self.observers { o.update(self.temperature) } }
    fn set_temperature(&mut self, temp: f32) { self.temperature = temp; self.notify() }
}

struct PhoneDisplay;
impl Observer for PhoneDisplay { fn update(&self, temp: f32) { println!("Phone: Temperature is {:.1}°C", temp) } }

struct WindowDisplay;
impl Observer for WindowDisplay { fn update(&self, temp: f32) { println!("Window: Temperature is {:.1}°C", temp) } }

fn main(){
    let mut station = WeatherStation::new();
    station.attach(Box::new(PhoneDisplay));
    station.attach(Box::new(WindowDisplay));
    station.set_temperature(25.5);
    station.set_temperature(30.2);
}
